[Merged by Bors] - feat: smooth immersions

grunweg opened the pull request:
Define smooth immersions between manifolds. We use a definition which allows for corners, infinite dimension and arbitrary fields. This matches the mathematical literature for e.g. real Banach manifolds with corners (even though there is no textbook treating manifolds as general as mathlib's).

This PR provides four central definitions: being an immersion at a point, and being an immersion everywhere,
plus a variant of these with a fixed complement. (Being an immersion at a point is a property with respect to some choice of complement in the model normed space of the codomain. In most applications, there is no need to control this complement. Such control will be helpful, however, when considering the local characterisation of submanifolds: locally, a submanifold is described either as the image of an immersion, or the preimage of a submersion --- w.r.t. the same complement. Having access to a definition version with complements allows stating this equivalence cleanly.

This PR only provides the basic API to keep its size (relatively) small. Future PRs will add more substance, and
- prove that immersions are C^n (#28796),
- prove another characterisation of immersions: f is immersed at x iff the differential df_x splits,
  i.e. is injective, has closed range and its image has a closed complement,
  and in particular that a map between finite-dimensional manifolds over a complete field is an
  immersion iff its differential is injective,
- use this to prove that the product (#28853) and composition of immersions is an immersion,
- use this theory to give more conceptual proofs of half of `Manifold/Instances/Icc.lean` (#29077)
- define smooth embeddings and immersed and embedded smooth submanifolds.

---

- [x] depends on: #28701
- [x] depends on: #31123
- [x] depends on: #31896
- [x] depends on: #31897
- [x] depends on: #31899

Most of the above has been done at https://github.com/grunweg/mathlib4/tree/MR-define-immersions.

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

grunweg committed:
chore(references.bib): add Roig and Domingues' book

grunweg committed:
change

grunweg committed:
Shake

grunweg committed:
Missing docs

grunweg committed:
Tweak

chrisflav commented on Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean:
```diff
@@ -0,0 +1,253 @@
+lemma isImmersionAt (h : IsImmersion F I I' n f) (x : M) : IsImmersionAt F I I' n f x := h x
+
+/-- If `f = g` and `f` is an immersion, so is `g`. -/
+theorem congr (h : IsImmersion F I I' n f) (heq : f = g) : IsImmersion F I I' n g :=
+  fun x ↦ (h x).congr_of_eventuallyEq heq.eventuallyEq
```
```suggestion
theorem congr (h : IsImmersion F I I' n f) (heq : f = g) : IsImmersion F I I' n g :=
  heq ▸ h
```

grunweg replied:
Much nicer, thanks!

chrisflav commented on Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean:
```diff
@@ -0,0 +1,253 @@
+NB. We don't know the particular atlasses used for `M` and `N`, so asking for `φ` and `ψ` to be
+in the `atlas` would be too optimistic: lying in the `maximalAtlas` is sufficient.
+-/
+def IsImmersionAt (f : M → M') (x : M) : Prop :=
+  ∃ equiv : (E × F) ≃L[𝕜] E',
+  ∃ domChart : PartialHomeomorph M H, ∃ codChart : PartialHomeomorph M' H',
+    x ∈ domChart.source ∧ f x ∈ codChart.source ∧
+    domChart ∈ IsManifold.maximalAtlas I n M ∧
+    codChart ∈ IsManifold.maximalAtlas I' n M' ∧
+    f '' domChart.source ⊆ codChart.source ∧
+    EqOn ((codChart.extend I') ∘ f ∘ (domChart.extend I).symm) (equiv ∘ (·, 0))
+      (domChart.extend I).target
```
Take my opinion with a big grain of salt, because I don't know the DG library at all: In AG, we would probably define this in the following way: Define a `IsSplitEmbeddingAt` which says gobally it is of the form `u \mapsto (u, 0)` (I suppose this would be a property of maps between subsets of model spaces) and then define `IsImmersionAt` as the source-local closure of this (the source-local closure of a property `P` (between subsets of model spaces) would be: "there exists a chart on the source and on the target such that the restriction satisfies `P`").
This helps with unifying proofs and separating concerns, because you can then start proving things about "if `P` is stable under composition, also `SourceLocalClosure P` is stable under composition" etc.

chrisflav replied:
I agree that showing "composition of `SourceTargetLocalClosure P` is `SourceTargetLocalClosure P`" is subtle, but I believe the proof will be much clearer if `P` is abstract. The key should be that `P` is stable under restrictions on the target.

In AG, it only rarely happens that a property is not local at the target, so we only have the closure on the source, but you can look at https://github.com/leanprover-community/mathlib4/blob/71061a6571d99c42ec22db81c3322453bc23cbe1/Mathlib/AlgebraicGeometry/Morphisms/LocalClosure.lean#L87-L92 to see the proof that `SourceLocalClosure P` is stable under composition if `P` is (and `P` is stable under base change, which would translate in your setting to being stable under restrictions to open sets on the target).

(I just realized that I wrote `SourceLocalClosure` in my first message, but it should probably be `SourceTargetLocalClosure`.)

grunweg committed:
Update Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean

Co-authored-by: Christian Merten <136261474+chrisflav@users.noreply.github.com>

sgouezel commented:
Nice, thanks!

grunweg committed:
refactor: introduce a notion of local properties,

and refactor immersions to use it

grunweg committed:
Generalise proof that continuity at x suffices for immersions

grunweg committed:
Generalise proof of congruence,
and add some basic API for local properties

grunweg commented:
@chrisflav I just pushed a first version using local properties to abstract some of the proofs. (They do indeed become a bit clearer.) What do you think?

grunweg committed:
Lint and hint at missing docs

grunweg committed:
Polish, and clearer name

grunweg committed:
Try better names

grunweg committed:
Add documentation

grunweg committed:
fixup: use immersion property more

chrisflav commented on Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean:
```diff
@@ -0,0 +1,386 @@
+
+Motivating examples are immersions and submersions of smooth manifolds. -/
+structure LocalSourceTargetPropertyAt (f : M → M') (x : M)
+    (P : (M → M') → PartialHomeomorph M H → PartialHomeomorph M' H' → Prop) where
```
Why does this take `f` and `x` as a parameter? I would expect this to have signature:
```
LocalSourceTargetPropertyAt : (P : (M → M') → PartialHomeomorph M H → PartialHomeomorph M' H' → Prop) → Prop
```

chrisflav replied:
Then it would be rather `IsLocalSourceTargetProperty`.

chrisflav commented on Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean:
```diff
@@ -0,0 +1,386 @@
+structure LocalSourceTargetPropertyAt (f : M → M') (x : M)
+    (P : (M → M') → PartialHomeomorph M H → PartialHomeomorph M' H' → Prop) where
+  mono_source : ∀ f : M → M', ∀ φ : PartialHomeomorph M H, ∀ ψ : PartialHomeomorph M' H',
+    ∀ s : Set M, P f φ ψ → P f (φ.restr s) ψ
```
I think you only need to require this for `IsOpen s`?

chrisflav reviewed (commented):
Thanks for trying out my suggestion! I generally like the way this looks, I left some questions and comments below. (When the precise design is figured out, all the general parts about local properties should go to a separate file of course.)

grunweg committed:
Weaken first property

grunweg committed:
Better signature of LocalSourceTargetPropertyAt, and rename

grunweg committed:
chore: move the equivalence around; is this better?

grunweg commented on Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean:
```diff
@@ -0,0 +1,386 @@
+in the `atlas` would be too optimistic: lying in the `maximalAtlas` is sufficient.
+-/
+def IsImmersionAt (f : M → M') (x : M) : Prop :=
+  ∃ equiv : (E × F) ≃L[𝕜] E',
+  LiftSourceTargetPropertyAt I I' n f x (ImmersionAtProp I I' equiv)
```
Good question! Let me think about this. (It certainly seems possible to do it your way.)

grunweg replied:
I think putting the equivalence inside `ImmersionAtProp` also makes sense conceptually.

grunweg committed:
chore: move everything about local properties to a new file

grunweg committed:
Tweak

grunweg committed:
Tweaks

chrisflav commented on Mathlib/Geometry/Manifold/LocalSourceTargetProperty.lean:
```diff
@@ -0,0 +1,165 @@
+
+## Main definitions and results
+
+* `LocalSourceTargetPropertyAt` captures a local property of the above form: for each `f: M → N`,
+  `x : M` and charts `φ` of `M` around `x` and `ψ` of `N` around `f x`, the local property is either
+  safisfied or not. We ask that the property be stable under restriction of `φ` and local near `x`.
```
This is slightly outdated.

chrisflav commented on Mathlib/Geometry/Manifold/LocalSourceTargetProperty.lean:
```diff
@@ -0,0 +1,165 @@
+lemma property (h : LiftSourceTargetPropertyAt I I' n f x P) : P f h.domChart h.codChart :=
+  (Classical.choose_spec (Classical.choose_spec h)).2.2.2.2.2
+
+/-- If `P` is monotone w.r.t. restricting `domChart`, then it suffices to prove continuity of `f`
+at `x` (instead of a relation between the chart's sources). -/
```
I don't understand this docstring. Do you mean "if f is continuous at x, it suffices to show `P f domChart codChart` for some choice of `domChart` and `codChart` not necessarily satisfying `f '' domChart.source ⊆ codChart.source`"?
(This is a bit of a complicated sentence, but I don't have a better idea right now.)

grunweg replied:
Good point. How do you like the current phrasing?

chrisflav commented on Mathlib/Geometry/Manifold/LocalSourceTargetProperty.lean:
```diff
@@ -0,0 +1,165 @@
+-/
+def LiftSourceTargetPropertyAt (f : M → M') (x : M)
+    (P : (M → M') → PartialHomeomorph M H → PartialHomeomorph M' H' → Prop) : Prop :=
+  ∃ domChart : PartialHomeomorph M H, ∃ codChart : PartialHomeomorph M' H',
+    x ∈ domChart.source ∧ f x ∈ codChart.source ∧
+    domChart ∈ IsManifold.maximalAtlas I n M ∧
+    codChart ∈ IsManifold.maximalAtlas I' n M' ∧
+    f '' domChart.source ⊆ codChart.source ∧
+    P f domChart codChart
```
You could consider making the data here a structure `LocalPresentationAt` and replace this by `Nonempty (LocalPresentationAt P f x)`. This would make the lines below less unreadable.

grunweg replied:
Good idea, also done. This makes the code longer, but hopefully easier to understand.

chrisflav commented on Mathlib/Geometry/Manifold/LocalSourceTargetProperty.lean:
```diff
@@ -0,0 +1,165 @@
+
+Motivating examples are immersions and submersions of smooth manifolds. -/
+structure IsLocalSourceTargetProperty
+    (P : (M → M') → PartialHomeomorph M H → PartialHomeomorph M' H' → Prop) : Prop where
```
The hard part is figuring out the precise conditions that should be required here, of course these can still be changed later or added as additional predicates on these properties.
What comes to my mind right now: What should happen on restriction on the target? Should we add a condition for this? Should `mono_source` be part of this or be a separate condition?

Probably, it is fine to leave this as is and take this as a first approximation. It will be easy to change later (this is what happened over and over again for the corresponding definitions in algebraic geometry / commutative algebra).

chrisflav commented on Mathlib/Geometry/Manifold/LocalSourceTargetProperty.lean:
```diff
@@ -0,0 +1,165 @@
+    (P : (M → M') → PartialHomeomorph M H → PartialHomeomorph M' H' → Prop) : Prop where
+  mono_source : ∀ f : M → M', ∀ φ : PartialHomeomorph M H, ∀ ψ : PartialHomeomorph M' H',
+    ∀ s : Set M, IsOpen s → P f φ ψ → P f (φ.restr s) ψ
+  congr : ∀ f g : M → M', ∀ φ : PartialHomeomorph M H, ∀ ψ : PartialHomeomorph M' H',
+    ∀ s : Set M, IsOpen s → EqOn f g s → P f (φ.restr s) ψ → P g (φ.restr s) ψ
```
Should this be stated for general `φ` with assumption `φ.source ⊆ s` instead of `φ.restr s`? Also, I would think an `iff` version of this would be useful as a lemma.

grunweg replied:
Good ideas, both of them!

chrisflav commented on Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean:
```diff
@@ -0,0 +1,256 @@
+
+omit [ChartedSpace H M] [ChartedSpace H' M'] in
+/-- Being an immersion at `x` is a "nice" local property. -/
+lemma ImmersionAtPropIsNice : IsLocalSourceTargetProperty (ImmersionAtProp F I I' M M') where
```
```suggestion
lemma isLocalSourceTargetProperty_immersionAtProp : IsLocalSourceTargetProperty (ImmersionAtProp F I I' M M') where
```

chrisflav commented on Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean:
```diff
@@ -0,0 +1,256 @@
+`f` will look like an inclusion `u ↦ (u, 0)` in these extended charts.
+The particular chart is arbitrary, but this choice matches the witnesses given by
+`h.equiv` and `h.domChart`. -/
+noncomputable def codChart (h : IsImmersionAt F I I' n f x) : PartialHomeomorph M' H' :=
+  Classical.choose (Classical.choose_spec h)
+
+lemma mem_domChart_source (h : IsImmersionAt F I I' n f x) : x ∈ h.domChart.source :=
+  (Classical.choose_spec (Classical.choose_spec h)).1
+
+lemma mem_codChart_source (h : IsImmersionAt F I I' n f x) : f x ∈ h.codChart.source :=
```
Are you keeping these to have an abstraction boundary? If yes, then at least you should use the `LiftSourceTargetPropertyAt.codChart` etc. definitions here.

grunweg replied:
Yes, that was my idea. (There are some lemmas about immersions whose statements require e.g. h.domChart.source.) Good point about the proofs!

grunweg committed:
Better docs

grunweg committed:
generalise congruence property

grunweg committed:
Add congr_iff lemma for local properties

grunweg committed:
Update Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean

Co-authored-by: Christian Merten <136261474+chrisflav@users.noreply.github.com>

grunweg committed:
Fixup

grunweg committed:
Better proofs of IsImmersionAt API fields

grunweg committed:
Comment

grunweg committed:
Extract local presentation data into a structure:
makes the proofs below much more readable

grunweg committed:
Fix

grunweg commented on Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean:
```diff
@@ -0,0 +1,258 @@
+    simp [Φ]
+
+variable (F I I' n) in
+/-- `f : M → N` is a `C^k` immersion at `x` if there are charts `φ` and `ψ` of `M` and `N`
+around `x` and `f x`, respectively such that in these charts, `f` looks like `u ↦ (u, 0)`.
+Additionally, we demand that `f` map `φ.source` into `ψ.source`.
```
Good point! Mathematically, I would be equally happy with `F` wrapped as an exists as well. I'm not sure how to make that work, though - given that `F` needs to have the typeclass hypotheses `[NormedAddCommGroup F] [NormedSpace 𝕜 F]` for the type of `equiv` to even make sense.

grunweg replied:
I have certainly tweaked the comment.

grunweg replied:
I'm thus inclined to go ahead with the current one, and add the latter version in case we ever need it.

grunweg replied:
(I also tried defining the F-independent version. It didn't look much nicer to me; I'd prefer not to have a universe parameter inside the definition of immersions by default. Not without a compelling argument that this is necessary, which at the moment I don't see.)

chrisflav replied:
> Yes and no. The codimension of a _submanifold_ matters, but less so for an immersion. So I'd argue this is the wrong place to make this distinction.

But you _are_ making a distinction, because `IsImmersionAt` is in fact `IsImmersionWithFixedComplement`. If you don't want to change the name, this should at least be mentioned in the docstring.

chrisflav replied:
> (I also tried defining the F-independent version. It didn't look much nicer to me; I'd prefer not to have a universe parameter inside the definition of immersions by default. Not without a compelling argument that this is necessary, which at the moment I don't see.)

If you want to define a `F`-independent version, it would IMO be `IsImmersionAt f x := \exists (F : Type _) (_ : NormedAddCommGroup F) ..., IsImmersionWithFixedComplement F f x`. I totally agree with you that your definition is good and useful as is and should be a component of the `F`-independent one.

grunweg replied:
Fair point; I added this to the doc-string of `IsImmersion(At)`.

grunweg committed:
Review comments

grunweg committed:
More tweaks

grunweg commented on Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean:
```diff
@@ -0,0 +1,258 @@
+   _ = Ψ '' Ψ.source := by rw [PartialHomeomorph.extend_source]
+   _ ⊆ _ := Ψ.map_source''
+
+/-- If `f` is an immersion at `x` and `g = f` on some neighbourhood of `x`,
+then `g` is an immersion at `x`. -/
+lemma congr_of_eventuallyEq {x : M} (hf : IsImmersionAt F I I' n f x) (hfg : f =ᶠ[nhds x] g) :
```
Good point. Sounds good, both of it. I'm leaving this comment open for future reference, but won't act on it in this PR.

grunweg commented:
Thanks again! I have addressed all comments.

grunweg committed:
try: make IsImmersionAt indep of F... not as nice!

grunweg committed:
Revert "try: make IsImmersionAt indep of F... not as nice!"

This reverts commit 9ea95cb63dbc9a6bcfa8fc0f094bf779e25b10f1.

grunweg committed:
Namespace under Manifold

grunweg commented:
Side note: trying your suggestion about an F-independent immersion fails, as follows:
```
-- put this right after the definition of `IsImmersionAt`
/--
error: Failed to infer universe levels in binder type
  NormedSpace.{u_1, ?u.7048} 𝕜 F
---
error: Failed to infer universe levels in binder type
  NormedAddCommGroup.{?u.7048} F
---
error: Failed to infer universe levels in type of binder `F`
  Type ?u.7048
-/
#guard_msgs in
def IsImmersionAt2 (f : M → M') (x : M) : Prop :=
  ∃ (F : Type _), ∃ (_ : NormedAddCommGroup F), ∃ (_ : NormedSpace 𝕜 F), IsImmersionAt F I I' n f x

grunweg committed:
One more comment

grunweg committed:
And one more

chrisflav reviewed (commented):
This is looking quite good now. I suggested some non-mathematical golfs below.

There are a few lemmas that I think should exist, but given that this PR is already quite long, they can probably wait, but could maybe be added as a TODO:
- Show `IsImmersion(At)` is stable under replacing `F` by an isomorphic copy.
- The set where `LiftSourceTargetPropertyAt` holds is open.

grunweg commented on Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean:
```diff
@@ -0,0 +1,262 @@
+/-- If `f` is an immersion, it is an immersion at each point. -/
+lemma isImmersionAt (h : IsImmersion F I I' n f) (x : M) : IsImmersionAt F I I' n f x := h x
+
+/-- If `f = g` and `f` is an immersion, so is `g`. -/
+theorem congr (h : IsImmersion F I I' n f) (heq : f = g) : IsImmersion F I I' n g :=
+  heq ▸ h
```
This lemma is by analogy with `IsImmersionAt.congr_of_eventuallyEq`... so I don't expect this one to be used much, but think it's nice to include for API completeness.

grunweg committed:
Docs, namespace and reword doc-string

grunweg commented on Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean:
```diff
@@ -0,0 +1,262 @@
+lemma property (h : IsImmersionAt F I I' n f x) :
+    LiftSourceTargetPropertyAt I I' n f x (ImmersionAtProp F I I' M M') := h
+
+/-- Roig and Domingues [roigdomingues1992] only require this condition on the local charts:
```
Reworded, how about the new version?

chrisflav replied:
I am still confused by what you mean with "require". Do you want to say "Under the assumptions of Roig and Domingues, this lemma only holds for the local charts?"

grunweg replied:
No, I mean: Roig and Domingues' definition asks for the inclusion between the targets.
Our definition is different - because we use the junk value pattern.

grunweg commented on Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean:
```diff
@@ -0,0 +1,262 @@
+noncomputable def equiv (h : IsImmersionAt F I I' n f x) : (E × F) ≃L[𝕜] E' :=
+  Classical.choose <| LiftSourceTargetPropertyAt.property h
+
+lemma writtenInCharts (h : IsImmersionAt F I I' n f x) :
```
The name `writtenInCharts` does not exist, but is by analogy to `writtenInExtCharts` (which is a term used in `ExtChartAt.lean` --- so I have a preference for keeping that part.
I could rename it to `writtenInCharts_eqOn`, though. WDYT?

grunweg committed:
Add IsImmersionAt.def

grunweg committed:
Golf

Co-authored-by: Christian Merten <136261474+chrisflav@users.noreply.github.com>

grunweg committed:
Golf

Co-authored-by: Christian Merten <136261474+chrisflav@users.noreply.github.com>

grunweg committed:
Golf

Co-authored-by: Christian Merten <136261474+chrisflav@users.noreply.github.com>

grunweg committed:
Review comments

grunweg commented:
@chrisflav Thanks a lot for the careful review, again! I addressed all of your comments (except about rephrasing the doc-string, which I'll do tonight).
@sgouezel This is reaching the point where the design feels stable, and this is ready for a look by a differential geometry expert. Would you like to take a look?

grunweg committed:
Shake

grunweg committed:
Reword doc-string again

sgouezel commented on Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean:
```diff
@@ -0,0 +1,255 @@
+    (hsource : f '' domChart.source ⊆ codChart.source)
+    (hwrittenInExtend : EqOn ((codChart.extend I') ∘ f ∘ (domChart.extend I).symm) (equiv ∘ (·, 0))
+      (domChart.extend I).target) : IsImmersionAt F I I' n f x := by
+  use domChart, codChart; use equiv
```
Use a single bracket term instead of `use`?

sgouezel commented on Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean:
```diff
@@ -0,0 +1,255 @@
+  IsImmersionAt.def.1 h
+
+/--
+Roig and Domingues' [roigdomingues1992] definition of immersions only asks for this inclusion
```
Ideally, a docstring should be self-contained, i.e., readable without looking at the lemma statement. Here it's not the case, because "this inclusion" is not explained.

grunweg committed:
Some review comments

grunweg committed:
Update Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

grunweg committed:
chore: use preimages instead of images

grunweg committed:
More review comments

grunweg committed:
Speak about a function M \to N all the time.

grunweg committed:
Make IsImmersionAt an irreducible_def

grunweg committed:
More review comments

grunweg committed:
chore: rename manifolds, preparations

grunweg committed:
Merge branch 'master' into def-immersions

grunweg committed:
Fix for merge

sgouezel reviewed (commented):
I'm not sure I'm convinced by the design where `F` is a parameter. Instead, in the same way that you have "there exists an equiv", why not put "there exists F"? Otherwise, I'm afraid that whenever we speak of immersions in finite dimension we will need to write everywhere `Fin (n - i) -> R` and this will make things harder to read. Do you have a mathematical reason for wanting to include `F`? Or a formalization reason?

grunweg committed:
Update Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

chrisflav commented:
> I'm not sure I'm convinced by the design where `F` is a parameter. Instead, in the same way that you have "there exists an equiv", why not put "there exists F"? Otherwise, I'm afraid that whenever we speak of immersions in finite dimension we will need to write everywhere `Fin (n - i) -> R` and this will make things harder to read. Do you have a mathematical reason for wanting to include `F`? Or a formalization reason?

I don't know if you have seen the thread of comments starting here: https://github.com/leanprover-community/mathlib4/pull/28793#discussion_r2397852779.

TLDR: I suggested to rename the notion defined in this PR to `IsImmersionAtOfComplement` (or something similar) and introduce an `IsImmersionAt := ∃ (F : Type _), ∃ (_ : NormedAddCommGroup F), ∃ (_ : NormedSpace 𝕜 F), IsImmersionAtOfComplement F I I' n f x`.
(If it helps, we have a docs#[AlgebraicGeometry.IsSmoothOfRelativeDimension](https://leanprover-community.github.io/mathlib4_docs/Mathlib/AlgebraicGeometry/Morphisms/Smooth.html#AlgebraicGeometry.IsSmoothOfRelativeDimension) and a [AlgebraicGeometry.IsSmooth](https://leanprover-community.github.io/mathlib4_docs/Mathlib/AlgebraicGeometry/Morphisms/Smooth.html#AlgebraicGeometry.IsSmooth) in AG and they are both useful, depending on the context.)

sgouezel commented:
I hadn't seen the discussion, thanks for the pointer. One rightful objection by Michael is that we shouldn't have a free universe parameter around, and I agree with that. So instead of your suggestion it should probably be
`∃ (F : Type u), ∃ (_ : NormedAddCommGroup F), ∃ (_ : NormedSpace 𝕜 F), IsImmersionAtOfComplement F I I' n f x`
where `u` is the universe of `E''`.
I'm fine with having both definitions (one with a parameter `F`, and one without) but I have the impression that the main one should be the one without `F` because it is less verbose and equally powerful. @grunweg, what are your feelings on this?
(No need to do everything in this PR, but we should settle on the design now, because this will influence the names we use)

chrisflav commented:
I totally agree with the universe parameter, I left it out for brevity. I am not sure what `E''` is in your suggestion, but I would assume you mean `u` should be the universe parameter of the model space of the target?

grunweg committed:
Simplify congruence property

grunweg commented on Mathlib/Geometry/Manifold/LocalSourceTargetProperty.lean:
```diff
@@ -0,0 +1,198 @@
+  -- Note: the analogous `mono_target` statement is true for both immersions and submersions.
+  -- If and when a future lemma requires it, add this here.
+  congr : ∀ {f g : M → N}, ∀ {φ : OpenPartialHomeomorph M H}, ∀ {ψ : OpenPartialHomeomorph N G},
+    ∀ {s : Set M}, EqOn f g s → IsOpen s → φ.source ⊆ s → P f φ ψ → P g φ ψ
```
Good point. Indeed, the latter should be simpler. Changed.

grunweg commented:
I experimented with this question again: #31108 shows that adding a definition of "IsImmersionAt" without a choice of complement works fine (even without much API about the complement version). Right now, I'm taking the universe parameter from E'' (which is the model normed space of the target). I could probably take it from the domain also; that needs a bit more care with the `variable`s at the top of the file.

How do you feel about the next steps?
- do you mind if we land this PR first (hammering out the shape of all the API lemmas), and then landing a polished version of #31108 as a follow-up?
- should we have API both for immersions with a fixed complement and without? If the latter, it's probably better to change this PR directly...

grunweg committed:
chore(OpenPartialHomeomorph): tag two lemmas with grind

Follow-up from #28793.

grunweg committed:
Merge branch 'master' into def-immersions

grunweg committed:
Review comments

grunweg committed:
Merge branch 'grind-openpartialhom' into def-immersions

grunweg committed:
Simplify grind invocations

grunweg commented:
I'll rework this PR accordingly. Thanks for the feedback!

chrisflav commented:
I don't understand your second point (which is certainly just because I am not a differential geometer), but let me make one remark on the first one:

In the finite dimensional case, we can define `IsImmersionAt.codimension` which is the codimension of an arbitrary choice of complement. This is well-defined, because the complements are unique up to isomorphism. Then we can provide API that from an assumption `h : IsImmersionAt _ _ _ _` gives a local presentation with complement `Fin (IsImmersionAt h.codimension) -> R`. I think this will give the exact same advantages as the `OfComplement (Fin k -> R)` version, without the additional parameter.

grunweg commented:
> Submanifolds can equivalently be given by immersions or submersions.

I agree with this statement *locally*, but not globally. That's what you mean, right?

grunweg committed:
Rename IsImmersionAt to IsImmersionAtOfComplement, and
add IsImmersionAt (with the same basic API).
Indicate the difference between them.

grunweg commented:
I have just renamed `IsImmersionAt` to `IsImmersionAtOfComplement`, added `IsImmersionAt` (with the same basic API) and fixed all the proofs. While this makes the diff substantially large, this PR is now ready for another look.

sgouezel commented on Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean:
```diff
@@ -0,0 +1,440 @@
+
+universe u
+variable {𝕜 : Type*} [NontriviallyNormedField 𝕜]
+  {E E' E''' : Type*} {E'' F F' : Type u} [NormedAddCommGroup E] [NormedSpace 𝕜 E]
```
There should be no universe constraint on `F`: in `IsImmersionOfComplement`, spaces could be in arbitrary universes but still you might want to take `F` in Type.

This means that slightly more work will be needed to deduce `IsImmersionAt` (with a complement in `Type u`, because you don't want the definition to be polymorphic on an additional useless universe) from `IsImmersionAtOfComplement` (with a complement in an arbitrary universe) -- need some API to say that if you have `IsImmersionAtOfComplement F` and a linear equiv between `F` and `F'` you get `IsImmersionAtOfComplement F'`.

grunweg replied:
Can you elaborate on the rest of your comment: do you prefer to make `IsImmersionAtOfComplement` take in a complement in any universe, but the complement in `IsImmersionAt` to live in e.g. the same universe as `E''`? (Or in `Type`?) Naively trying this (as I just pushed) breaks two proofs, and I don't immediately see how to restore them.

sgouezel replied:
Yes, in `ImmersionAtOfComplement` the space `F` should be an arbitrary universe. While in `ImmersionAt` the space `F` should be taken in a universe coming from another object (for instance that of `E''`) to avoid having an additional universe parameter in the definition.

grunweg replied:
Makes sense. This matches what I currently have. Let me look harder at the two broken proofs...

chrisflav replied:
I think that mathematically, the universe of `F` must be the universe of `E''` in `ImmersionAt`: We should have a lemma that takes `ImmersionAtComplement F` as assumption and concludes: `Small.{u} F` where `u` is the universe parameter of `E''` (proof: use the linear equivalence of `E''` to the product of `E'` and `F`). Then we should have a constructor of `ImmersionAt` taking an `F` in an arbitrary universe and a proof of `ImmersionAtComplement F` (the proof is: take the small model of `F` in `Type u`).

chrisflav replied:
Maybe all of this (including `ImmersionAt` / `Immersion`) should be done in a follow-up PR though? This PR is already getting quite long.

grunweg committed:
Add both IsImmersion versions

grunweg committed:
Better IsImmersion definition: one sorry and some API to go

grunweg committed:
Docs

grunweg committed:
chore: generalise universes in IsImmersionOfComplement definition, WIP

grunweg committed:
More API about complements

grunweg committed:
Outline one proof: needs more mfderiv API, follow-up?

grunweg committed:
Note TODOs for more general universes

grunweg committed:
Remove WIP proof about immersion vs immersion at: needs more API first

grunweg committed:
Merge branch 'master' into def-immersions

grunweg committed:
close to a nice proof!

grunweg committed:
wip

grunweg committed:
Smallness lemmas complete

grunweg committed:
Almost there!

grunweg committed:
Further tinkering

grunweg committed:
Sorry-free; clean-up tomorrow

grunweg committed:
Clean up a bit

grunweg committed:
Some API lemmas and clean-up

sgouezel commented on Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean:
```diff
@@ -0,0 +1,578 @@
+* `IsImmersion I J n f` means `f : M → N` is an immersion at every point `x : M`,
+  w.r.t. some global choice of complement.
+  `isImmersion_iff_isImmersionAt` will prove this is equivalent to `f` being an
+  immersion at every point `x` (i.e., with respect to a complement that can depend on `x`).
```
Remove this for now. (I'm not even sure this is true...)

chrisflav replied:
Sebastien's comment makes sense to me, which means that the current definition of `IsImmersion` is probably mathematically not what we want and should be replaced by `\forall, IsImmersionAt` (which means the complement may depend on the point).

grunweg committed:
Final rename

grunweg committed:
Move lemmas; delete non-CLE versions

grunweg commented on Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean:
```diff
@@ -0,0 +1,585 @@
+as the model space for the co-domain of `f`: this is useful to avoid universe restrictions. -/
+def smallComplement (hf : IsImmersionAtOfComplement F I J n f x) : Type u :=
+  LinearMap.range (hf.equiv.domRestrict (LinearMap.range (LinearMap.prod 0 .id)))
+
```
Typing out a sketch would be very welcome! I'll have to prepare my teaching now, so won't get to it before tonight.

grunweg committed:
Merge branch 'master' into def-immersions

grunweg commented on Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean:
```diff
@@ -0,0 +1,564 @@
+-- Note: without completeness assumptions, one can still find a `LinearEquiv`:
+-- we don't provide this definition as it won't be used in practice.
+def smallEquiv [CompleteSpace E] [CompleteSpace E''] [CompleteSpace F]
+    (hf : IsImmersionAtOfComplement F I J n f x) : F ≃L[𝕜] hf.smallComplement := by
```
Note to self: [LinearEquiv.submoduleMap](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Module/Submodule/Map.html#LinearEquiv.submoduleMap) does something similar for LinearEquiv's, I want a continuous version.

grunweg committed:
Merge branch 'master' into def-immersions

grunweg committed:
Better proof, without helper lemmas

grunweg committed:
Helper lemmas, part 1

grunweg committed:
Last API for new definitions

grunweg committed:
Tweak

grunweg committed:
feat: transfer normed spaces and continuous linear equiv's across equivalences

Analogous to existing instances for e.g. linear equivalences.
This is required for adding Shrink instances for these, which in turns are
necessary for making the definition of immersions in #28793
universe-polymorphic.

grunweg committed:
Fix CI; better documentation

grunweg committed:
feat: add Shrink instances for NormedSpace and ContinuousLinearEquiv

grunweg committed:
feat: Shrink instances for NormedSpace and ContinuousLinearEquiv

grunweg committed:
Make public

grunweg committed:
Tweak and improve previous PR:
fix variable explicitness, generalise to semirings and fix variable order

grunweg committed:
Tweak and improve previous PR:
fix variable explicitness, generalise to semirings and fix variable order

grunweg commented:
#31897 adds the Shrink instances that were mentioned above.

grunweg committed:
Merge branch 'shrinknorm' into def-immersions

grunweg committed:
mk_all

grunweg committed:
feat: ContinuousLinearEquiv.submoduleMap

The continuous analogue of LinearEquiv.submoduleMap.

grunweg committed:
Also define ofSubmodules and submoduleMap'

grunweg committed:
Re-order: variable management is slightly annoying

grunweg committed:
Sorry-free: need to move the homeomorphism lemma

grunweg committed:
Move lemmas

grunweg committed:
Better name

grunweg committed:
Merge branch 'cle-submodulemap' into def-immersions

grunweg committed:
Remove now-duplicate definitions

grunweg committed:
Tried using the shrink instances: not set up right yet

grunweg committed:
Mostly replace the complements by a construction avoiding Shrink: one last sorry to solve

grunweg committed:
tinkering

grunweg committed:
Revert "tinkering"

This reverts commit 6136b0ef9b72b545704cf076d0fa7db05ebaf6c8.

chrisflav committed:
fix

chrisflav committed:
fix diamond: the old transported instances was .coinduced e.symm, cleanup

grunweg committed:
Merge branch 'shrinknorm' into def-immersions

grunweg committed:
It seems there is a diamond, right?

grunweg committed:
Does this fix the diamond? Need help with a last proof, though...

grunweg committed:
This doesn't seem to help either

chrisflav committed:
fix

chrisflav committed:
cleanup

grunweg committed:
Last clean-up

grunweg committed:
Merge branch 'shrinknorm' into def-immersions

grunweg committed:
Three plus one sorry left for the shrink route to small equivs

chrisflav committed:
fill sorries and cleanup

grunweg committed:
Merge branch 'master' into def-immersions

grunweg committed:
Merge branch 'master' into def-immersions

grunweg committed:
Fix merge: delete superfluous file

grunweg committed:
Adapt to module system, and fix the build

grunweg committed:
Review comments and linter

grunweg committed:
Note why an intended generalisation is false; fix typo.

And remove a stale commented line of code.

grunweg committed:
Small polish

grunweg committed:
Add second target_subset_preimage_target lemma, and more comments

grunweg commented:
I just re-read the entire file again: from my point of view, this would be good to go. What do you think, Christian and Sébastien?

chrisflav commented on Mathlib/Geometry/Manifold/LocalSourceTargetProperty.lean:
```diff
@@ -0,0 +1,200 @@
+variable (I I' n) in
+/-- A property of smooth functions `M → N` which is local at both the source and target:
+a property `P` is local at `x` iff there exist charts `φ` and `ψ` of `M` and `N` around
+`x` and `f x`, respectively, such that `f` satisfies the property w.r.t. `φ` and `ψ`.
```
This docstring does not really match the meaning here. Maybe something like this is clearer:
```suggestion
/-- The induced property by a local property `P`: It is satisfied for `f` at `x` iff there exist charts
`φ` and `ψ` of `M` and `N` around `x` and `f x`, respectively, such that `f` satisfies `P`
w.r.t. `φ` and `ψ`.
```

chrisflav reviewed (approved):
I went over it again. There are still some API lemmas missing here and there, but I think this is already a good start. So LGTM!

Not going to maintainer delegate, because probably @sgouezel wants to have another look.

grunweg committed:
Review comments

grunweg committed:
Different proof

grunweg committed:
Better doc-string

grunweg committed:
Remove misleading sentence

sgouezel commented on Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean:
```diff
@@ -0,0 +1,573 @@
+/-- Given an immersion `f` at `x`, this is a choice of complement which lives in the same universe
+as the model space for the co-domain of `f`: this is useful to avoid universe restrictions. -/
+def smallComplement (hf : IsImmersionAtOfComplement F I J n f x) : Type u :=
+  have := hf.small
```
```suggestion
  haveI := hf.small
```
(You can have a look at the generated term, it is nicer like that). Same thing below, in all the definitions and instances I'd replace `have := hf.small` with `haveI` to get nicer terms.

grunweg replied:
I know this in principle; thanks for reminding me. I'm not sure if it's as important in the instances, but I replaced those anyway. I strongly agree on the definitions.

sgouezel reviewed (commented):
Have you included the lemma saying that `IsImmersionAtOfComplement` implies `IsImmersionAt`? I thought it was the goal of all the `small` business, but I can't find it.

grunweg committed:
Use haveI

grunweg committed:
Missing API lemma

grunweg committed:
Missing lemma, modulo one sorry

grunweg committed:
Slightly further

grunweg committed:
Remove the empty case for now

grunweg committed:
Rename file: embeddings should go into a new file,

this one is/will be large enough already

grunweg committed:
Actually, the empty case is fine

grunweg committed:
Add implementation note

grunweg commented:
I just add two paragraphs of implementation notes and augmented the PR description.

sgouezel commented:
Great. Let's get this in. Thanks a lot for all the time you spent on this, and bearing with all our comments!
bors r+
